TombEngine/Objects/TR1/Trap/ElectricBall.cpp

```diff
@@ -0,0 +1,463 @@
+#include "Game/effects/item_fx.h"
+#include "Game/misc.h"
+#include "Objects/Effects/Boss.h"
+#include "Game/effects/spark.h"
+
+using namespace TEN::Math;
+using namespace TEN::Effects::Electricity;
+using namespace TEN::Effects::Items;
+using namespace TEN::Effects::Spark;
+using namespace TEN::Effects::Boss;
+
+namespace TEN::Entities::Traps
+{
+	enum TargetType
+	{
+		None,
+		Target
+	};
+
+	struct ElectricBallInfo
+	{
+		Vector3i Position = Vector3i::Zero;
+		Color Colorw;
+		Electricity* EnergyArcs = {};
+		Electricity* MainEnergyArc = {};
+		unsigned int Count = 0;
+	};
+
+	ElectricBallInfo EldectricBallData;
+
+	constexpr auto ELECTRIC_BALL_LIGHTNING_DAMAGE = 40;
+	const auto ElectricBallBite = CreatureBiteInfo(Vector3::Zero, 0);
+
+	 void TriggerElectricBallShockwaveAttackSparks(int x, int y, int z, byte r, byte g, byte b, byte size)
+	{
+		auto* spark = GetFreeParticle();
+
+		spark->dG = g;
+		spark->sG = g;
+		spark->colFadeSpeed = 2;
+		spark->dR = r;
+		spark->sR = r;
+		spark->blendMode = BlendMode::Additive;
+		spark->life = 4;
+		spark->sLife = 4;
+		spark->x = x;
+		spark->on = 1;
+		spark->dB = b;
+		spark->sB = b;
+		spark->fadeToBlack = 4;
+		spark->y = y;
+		spark->z = z;
+		spark->zVel = 0;
+		spark->yVel = 0;
+		spark->xVel = 0;
+		spark->flags = SP_SCALE | SP_DEF;
+		spark->scalar = 3;
+		spark->maxYvel = 0;
+		spark->spriteIndex = Objects[ID_DEFAULT_SPRITES].meshIndex + SPR_LENSFLARE_LIGHT;
+		spark->gravity = 0;
+		spark->dSize = spark->sSize = spark->size = size + (GetRandomControl() & 3);
+	}
+
+	const void SpawnElectricSmoke(const Vector3& pos, const EulerAngles& orient, float life)
+	{
+		auto& smoke = *GetFreeParticle();
+
+		float scale = (life * 18) * phd_cos(orient.x);
+
+		smoke.on = true;
+		smoke.blendMode = BlendMode::Additive;
+		smoke.spriteIndex = Objects[ID_DEFAULT_SPRITES].meshIndex + SPR_LENSFLARE_LIGHT;// SPR_LENSFLARE_LIGHT;
+
+		smoke.x = pos.x;
+		smoke.y = pos.y;
+		smoke.z = pos.z;
+		smoke.xVel = Random::GenerateInt(-25, 25);
+		smoke.yVel = life * 3;
+		smoke.zVel = Random::GenerateInt(-25, 25);
+		smoke.sB = (Random::GenerateInt(28, 196) * life) / 16;
+		smoke.sR =
+			smoke.sG = smoke.sB - (smoke.sB / 4);
+		
+		smoke.dB = (Random::GenerateInt(62, 80) * life) / 16;
+		smoke.dR = smoke.dB / 3;
+		smoke.dG = 
+			smoke.dB /2 ;
+
+		smoke.colFadeSpeed = Random::GenerateInt(8, 12);
+		smoke.fadeToBlack = 8;
+		smoke.sLife =
+			smoke.life = Random::GenerateFloat(24.0f, 28.0f);
+
+		smoke.friction = 0;
+		smoke.rotAng = Random::GenerateAngle(0, ANGLE(22.5f));
+		smoke.rotAdd = Random::GenerateAngle(ANGLE(-0.35f), ANGLE(0.35f));
+		smoke.gravity = Random::GenerateAngle(ANGLE(0.175f), ANGLE(0.35f));
+		smoke.maxYvel = 0;
+		smoke.scalar = 1;
+		smoke.size =
+			smoke.sSize = scale;
+		smoke.dSize = 1;
+		smoke.flags = SP_SCALE | SP_DEF | SP_ROTATE | SP_EXPDEF;
+	}
+
+	void InitializeElectricBall(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+		CheckForRequiredObjects(item);
```

Comment:
This should only be used for TR3 boss, GetTargetEntityList already do the job.
using CheckForRequiredObjects and changing the enum, you change file is not required in this PR.